### PR TITLE
Remove LeafMap - have a generic function which converts a Topology to RenderableNodes

### PR DESCRIPTION
**Pull request by tomwilkie**
Includes #421, those changes will go away when thats merged.

Part of #357 

Doesn't actually buy us as much as I'd hoped.

### CODE REVIEW

File: render/renderable_node.go [hunk 1]
```diff
@@ -107,6 +70,73 @@ func newDerivedPseudoNode(id, major string, node RenderableNode) RenderableNode
 		Pseudo:       true,
 		Origins:      node.Origins.Copy(),
 		EdgeMetadata: node.EdgeMetadata.Copy(),
-		NodeMetadata: report.MakeNodeMetadata(),
+		NodeMetadata: node.NodeMetadata.Copy(),
+	}
+}
+
+// WithNode creates a new RenderableNode bases on rn, with n
+func (rn RenderableNode) WithNode(n report.NodeMetadata) RenderableNode {
```
Comment (peterbourgon) [inline render/renderable_node.go]: It's not WithNode, but rather WithNodeMetadata...

**Comment by peterbourgon** [comment]
I appreciate the change. Removing the distinction of a LeafMapRenderer is a big win for understandability.

File: render/expected/expected.go [hunk 2]
```diff
@@ -10,22 +10,14 @@ import (
...
+	// state not sent to the client.  So in the tests we ignore
+	// this state.
 	for id, n := range r {
-		if n.Adjacency == nil {
-			n.Adjacency = report.MakeIDList()
-		}
-		n.NodeMetadata = report.MakeNodeMetadata()
+		n.NodeMetadata.Metadata = map[string]string{}
+		n.NodeMetadata.Counters = map[string]int{}
+		n.NodeMetadata.Edges = report.EdgeMetadatas{}
```
Comment (peterbourgon) [inline render/expected/expected.go]: I guess use the constructors here...

File: render/renderable_node.go (hunk 1 above)
Comment (tomwilkie) [inline render/renderable_node.go]: Sorry go ahead of myself; next change it to rename that struct.  Will fix.

**Comment by tomwilkie** [comment]
@peterbourgon PTAL

File: render/mapping.go [hunk 3]
```diff
@@ -25,26 +25,17 @@ const (
...
-// ID to a human-readable major and minor labels.
-//
-// A single NodeMetadata can yield arbitrary many representations, including
-// representations that reduce (or even increase) the cardinality of the set of nodes.
-type LeafMapFunc func(report.NodeMetadata, report.Networks) RenderableNodes
-
 // MapFunc is anything which can take an arbitrary RenderableNode and
 // return a set of other RenderableNodes.
 //
 // As with LeafMapFunc, if the output is empty, the node
```
Comment (paulbellamy) [inline render/mapping.go]: "As with LeafMapFunc"

File: render/renderable_node.go [hunk 4]
```diff
@@ -13,67 +13,42 @@ type RenderableNode struct {
 	LabelMinor string        `json:"label_minor,omitempty"` // e.g. "hostname", human-readable, optional
 	Rank       string        `json:"rank"`                  // to help the layout engine
 	Pseudo     bool          `json:"pseudo,omitempty"`      // sort-of a placeholder node, for rendering purposes
-	Adjacency  report.IDList `json:"adjacency,omitempty"`   // Node IDs (in the same topology domain)
 	Origins    report.IDList `json:"origins,omitempty"`     // Core node IDs that contributed information
 
 	report.EdgeMetadata `json:"metadata"` // Numeric sums
-	report.NodeMetadata `json:"-"`        // merged NodeMetadata of the nodes used to build this
+	report.NodeMetadata
```
Comment (peterbourgon) [inline render/renderable_node.go]: The non-omitted fields of NodeMetadata will get flattened and rendered into the RenderableNode JSON, i.e. you'll get

```json
{
    "label_minor": "whatever",
    "origins": ["a", "b", "c"],
    "adjacency": ["x", "y", "z"]
}
```

Is that your intent?

File: render/renderable_node.go [hunk 5]
```diff
@@ -107,6 +70,73 @@ func newDerivedPseudoNode(id, major string, node RenderableNode) RenderableNode
...
 		EdgeMetadata: node.EdgeMetadata.Copy(),
-		NodeMetadata: report.MakeNodeMetadata(),
+		NodeMetadata: node.NodeMetadata.Copy(),
+	}
+}
+
+// WithNodeMetadata creates a new RenderableNode based on rn, with n
+func (rn RenderableNode) WithNodeMetadata(n report.NodeMetadata) RenderableNode {
+	result := rn.Copy()
+	result.NodeMetadata = n
```
Comment (peterbourgon) [inline render/renderable_node.go]: Should this be

```go
result.NodeMetadata = result.NodeMetadata.Merge(n)
```

?

File: render/renderable_node.go (hunk 4 above)
Comment (tomwilkie) [inline render/renderable_node.go]: Yes thats my intention; NodeMetadata only exposes adjacency on json rendering.

File: render/renderable_node.go (hunk 5 above)
Comment (tomwilkie) [inline render/renderable_node.go]: Yes why not.  Its only used in constructors, so it should affect anything, but this would be nicer.

**Comment by peterbourgon** [comment]
:+1: otherwise.

Semi-related, I'm noticing that most churn in recent PRs is in the tests. I feel this myself, when I make a simple change in e.g. package render, I find I have to change a lot of "distant" tests in other components. I'm wondering if there's not a better testing strategy, relying a lot less or even even not at all on coarse-grained transforms of reports (i.e. package expected), and a lot more on fine-grained unit testing of individual functions and packages. What do you think?

**Comment by tomwilkie** [comment]
On one hand I agree, the tests are pretty coarse.  But at the end of the day they are a good way to catch unintended side effects of small changes that more focused test might miss, and they do provide excellent coverage.

An alternative would be to capture a json report and the json output of the renders and plumb them together; that might help with some of the tedious struct munging.   Or just make it tedious json munging instead.

**Pull request merged**